File: backend/app/services/pricing.py
"""Calculo do preco total entregue (secao 12).

Preco total imediato = preco + frete + impostos + taxas
                       - desconto no Pix - cupom valido.
Cashback NUNCA e abatido: aparece como beneficio posterior.
"""

from app.schemas.models import Offer, PriceBreakdown


def compute_price_breakdown(offer: Offer) -> PriceBreakdown:
    price = offer.price
    price_pix = offer.price_pix if offer.price_pix and offer.price_pix < price else price
    pix_discount = round(price - price_pix, 2)

    coupon_discount = 0.0
    if offer.coupon and offer.coupon.validated:
        base_for_coupon = price_pix
        if base_for_coupon >= offer.coupon.min_order_value:
            coupon_discount = min(offer.coupon.discount_value, base_for_coupon)

    shipping = offer.shipping_cost or 0.0
    taxes = offer.taxes or 0.0
    fees = offer.fees or 0.0
    if offer.import_cost:
        shipping = shipping or offer.import_cost.international_shipping
        taxes = taxes or (offer.import_cost.taxes_included + offer.import_cost.taxes_estimated)
        fees = fees or offer.import_cost.fees

    total = round(price_pix + shipping + taxes + fees - coupon_discount, 2)
    cashback_later = round(offer.cashback.value, 2) if offer.cashback else 0.0

    return PriceBreakdown(
        price=price,
        price_pix=price_pix,
        pix_discount=pix_discount,
        coupon_discount=coupon_discount,
        shipping=shipping,
        taxes=taxes,
        fees=fees,
        total_delivered=total,
        cashback_later=cashback_later,
    )

File: frontend/src/lib/demo/pricing.ts
// Preco total entregue (porta de backend/app/services/pricing.py).
// Cashback NUNCA e abatido do preco imediato.
import type { Offer, PriceBreakdown } from "../types";

type RawOffer = Offer & {
  taxes?: number;
  fees?: number;
  import_cost?: {
    international_shipping: number;
    taxes_included: number;
    taxes_estimated: number;
    fees: number;
    risk_of_extra_charges: boolean;
  } | null;
  coupon: (Offer["coupon"] & { min_order_value?: number }) | null;
};

export function computeBreakdown(offer: RawOffer): PriceBreakdown {
  const price = offer.price;
  const pricePix = offer.price_pix && offer.price_pix < price ? offer.price_pix : price;
  const pixDiscount = round2(price - pricePix);

  let couponDiscount = 0;
  if (offer.coupon?.validated) {
    const min = offer.coupon.min_order_value ?? 0;
    if (pricePix >= min) couponDiscount = Math.min(offer.coupon.discount_value, pricePix);
  }

  let shipping = offer.shipping_cost ?? 0;
  let taxes = offer.taxes ?? 0;
  let fees = offer.fees ?? 0;
  if (offer.import_cost) {
    shipping = shipping || offer.import_cost.international_shipping;
    taxes = taxes || offer.import_cost.taxes_included + offer.import_cost.taxes_estimated;
    fees = fees || offer.import_cost.fees;
  }

  return {
    price,
    price_pix: pricePix,
    pix_discount: pixDiscount,
    coupon_discount: couponDiscount,
    shipping,
    taxes,
    fees,
    total_delivered: round2(pricePix + shipping + taxes + fees - couponDiscount),
    cashback_later: round2(offer.cashback?.value ?? 0),
  };
}

function round2(v: number): number {
  return Math.round(v * 100) / 100;
}
